--- sources/wine/dlls/ntdll/unix/loader.c	2026-05-01 10:21:13
+++ sources/wine/dlls/ntdll/unix/loader.c	2026-05-01 10:21:46
@@ -651,8 +651,9 @@
 #endif
 
 
-static void preloader_exec( char **argv, const char *image_path )
+static void preloader_exec( char **argv, const char *image_path, WORD machine )
 {
+    char *path;
 #ifdef HAVE_WINE_PRELOADER
     asprintf( &argv[0], "%s-preloader", argv[1] );
 #ifdef __APPLE__
@@ -683,7 +684,12 @@
         replace_wineloader_path_with_link( &(argv[1]), image_path );
 #endif
 
-    execv( argv[1], argv + 1 );
+    path = getenv( "ROSETTA_X87_PATH" );
+    if (path && machine == IMAGE_FILE_MACHINE_I386) {
+        argv[0] = strdup( path );
+        execv( argv[0], argv );
+    } else
+        execv( argv[1], argv + 1 );
 }
 
 /* exec the appropriate wine loader for the specified machine */
@@ -693,10 +699,10 @@
 
     putenv( noexec );
 
-    if (((argv[1] = get_alternate_wineloader( machine )))) preloader_exec( argv, image_path );
+    // if (((argv[1] = get_alternate_wineloader( machine )))) preloader_exec( argv, image_path, machine );
 
     argv[1] = strdup( wineloader );
-    preloader_exec( argv, image_path );
+    preloader_exec( argv, image_path, machine );
     return STATUS_INVALID_IMAGE_FORMAT;
 }
 
@@ -2541,7 +2547,7 @@
     }
 
     /* default to 32-bit loader to support 32-bit prefixes */
-    if (machine == IMAGE_FILE_MACHINE_AMD64) machine = IMAGE_FILE_MACHINE_I386;
+    // if (machine == IMAGE_FILE_MACHINE_AMD64) machine = IMAGE_FILE_MACHINE_I386;
 
     loader_exec( new_argv, machine, argv[0] );
     fatal_error( "could not exec the wine loader\n" );
